### PR TITLE
WIP: Feat/remove bluebird

### PR DESCRIPTION
**Pull request by cliedeman**
Pre emptive PR for when the version of knex is released which drops node 6 and removes bluebird.

~~Another open question is how to deal with these methods in QueryBuilder:~~
I have just marked these as deprecated for now and do a bluebird wrap inside them

```
  then(...args) {
    const promise = this.execute();
    return promise.then(...args);
  }

  map(...args) {
    const promise = this.execute();
    return promise.map(...args);
  }

  reduce(...args) {
    const promise = this.execute();
    return promise.reduce(...args);
  }

  catch(...args) {
    const promise = this.execute();
    return promise.catch(...args);
  }

  return(...args) {
    const promise = this.execute();
    return promise.return(...args);
  }

  reflect() {
    const promise = this.execute();
    return promise.reflect();
  }

  bind(...args) {
    const promise = this.execute();
    return promise.bind(...args);
  }

  asCallback(...args) {
    const promise = this.execute();
    return promise.asCallback(...args);
  }

  nodeify(...args) {
    const promise = this.execute();
    return promise.nodeify(...args);
  }
```

~The overarching goal here is to enable [better stack traces natively](https://v8.dev/blog/v8-release-72#async-stack-traces) but that would require a lot more changes. Removing all Promise.resolve etc calls and adding async await in more places.~ Think this is mostly done.

Summary
* Renamed bluebird import to `const Bluebird = require('bluebird');` to make usage obvious
* Removed all uses of bluebird in non test code except for the methods above
* Hard to remove bluebird usage from test code due to lots of `delay` and `timeout` calls, @koskimas is happy to keep bluebird as a dev dependency (After the functions above have been dealt with.) 
* Added deprecation warnings to bluebird wrappers. map, reduce, catch etc.

Ciaran

### CODE REVIEW

**Comment by kibertoad** [comment]
As a matter of fact, knex already did in master. We could you some help with weedng out leftovers, though 😅.

**Comment by koskimas** [comment]
Thanks for working on this! You need to target the v2.0 branch though since objection follows semver and this is a breaking change.

**Comment by koskimas** [comment]
Now you are trying to merge v2.0 to master :smile:

**Comment by cliedeman** [comment]
I rebased earlier. will open a new PR soon. Can't seem to edit the target

**Comment by cliedeman** [comment]
Progress...

```
  1572 passing (19s)
  3 pending
  2 failing

  1) QueryBuilder
       should have knex methods:
     Error: knex method 'clearHaving' is missing from QueryBuilder
      at Assertion.assert (node_modules/expect.js/index.js:96:13)
      at Assertion.fail (node_modules/expect.js/index.js:489:10)
      at Context.it (tests/unit/queryBuilder/QueryBuilder.js:95:23)

  2) QueryBuilder
       should pass the query builder as `this` and parameter for the hooks:
     Error: abort
      at QueryBuilder.forClass.runBefore.onBuild.onBuildKnex.runAfter.runAfter.runAfter (tests/unit/queryBuilder/QueryBuilder.js:653:15)
      at RunAfterOperation.onAfter3 (lib/queryBuilder/operations/RunAfterOperation.js:17:22)
      at promiseUtils.try (lib/queryBuilder/QueryBuilderOperationSupport.js:360:37)
      at Object.promiseTry [as try] (lib/utils/promiseUtils/try.js:8:26)
      at QueryBuilder.callAsyncOperationMethod (lib/queryBuilder/QueryBuilderOperationSupport.js:360:11)
      at promise.then.result (lib/queryBuilder/QueryBuilder.js:1360:19)
```

The clearHaving test is failing on v2.0 as well.
Not sure how to fix ```should pass the query builder as `this` and parameter for the hooks:```

**Comment by koskimas** [comment]
@cliedeman Could you rebase again. The CI was borked, but it's now fixed.

**Comment by cliedeman** [comment]
@koskimas v2.0 still looks broken?

**Comment by koskimas** [comment]
Oh crap, that's true. I'll cherry-pick the fix from master to v2

File: lib/transaction.js [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
 const promiseUtils = require('./utils/promiseUtils');
 const { isSubclassOf } = require('./utils/classUtils');
 const { isFunction } = require('./utils/objectUtils');
 
-function transaction() {
+async function transaction() {
   // There must be at least one model class and the callback.
   if (arguments.length < 2) {
-    return Bluebird.reject(
+    return Promise.reject(
```
Comment (cliedeman) [inline lib/transaction.js]: can probably change all these `Promise.reject(new Error(...` to straight throws inside the async methods

File: testUtils/TestSession.js [hunk 2]
```diff
@@ -432,7 +433,7 @@ function inc(obj, key) {
 }
 
 function registerUnhandledRejectionHandler() {
-  Promise.onPossiblyUnhandledRejection(error => {
+  process.on('unhandledRejection', error => {
```
Comment (cliedeman) [inline testUtils/TestSession.js]: This is not a 100% proper substitute

**Comment by koskimas** [comment]
I've rebased v2.0 on top of master. There shouldn't be that many conflicts if you just rebase on top of the new v2 and drop the duplicate commits. The duplicate commits are conflicting atm.

**Comment by cliedeman** [comment]
Rebased.  a few new failures I will dig into but not sure when I will get to it

File: lib/queryBuilder/QueryBuilder.js [hunk 3]
```diff
@@ -630,43 +630,63 @@ class QueryBuilder extends QueryBuilderBase {
   }
 
   map(...args) {
+    deprecate(
+      'map is deprecated. Use the modify method instead. applyModifier will be removed in objection 3.0'
```
Comment (cliedeman) [inline lib/queryBuilder/QueryBuilder.js]: comment is wrong. Should be: map is deprecated. map will be removed in objection 3.0

File: lib/queryBuilder/QueryBuilder.js [hunk 4]
```diff
@@ -630,43 +630,63 @@ class QueryBuilder extends QueryBuilderBase {
...
+    deprecate('asCallback is deprecated. asCallback will be removed in objection 3.0');
+
     const promise = this.execute();
-    return promise.asCallback(...args);
+    return Bluebird.resolve(promise).asCallback(...args);
   }
 
   nodeify(...args) {
+    deprecate(
+      'bind is nodeify. Use the modify method instead. applyModifier will be removed in objection 3.0'
```
Comment (cliedeman) [inline lib/queryBuilder/QueryBuilder.js]: comment is wrong

**Comment by koskimas** [comment]
Thanks again for working on this. I can take it from here.

**Comment by kibertoad** [comment]
Maybe bluebird and pokyfills could be remived and min knex version bumped then?

**Pull request merged**